### PR TITLE
StreamLogFile fixes and improvements.

### PR DESCRIPTION
**Pull request by WenbinZhu**
## Overview

Fixed some defects and did some minor performance improvements.

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 1]
```diff
@@ -1064,13 +1061,7 @@ public void append(List<LogData> range) {
 
         // Extract all addresses associated with the provided write range.
         Set<Long> pendingWrites = range.stream()
-                .map(entry -> entry.getGlobalAddress()).collect(Collectors.toSet());
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Why remove the check?

Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Because the verify() method already did the check for uniqueness.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 2]
```diff
@@ -366,7 +363,7 @@ private void trimPrefix() {
         // (i.e. trim only complete segments)
         long endSegment = dataStore.getStartingAddress() / RECORDS_PER_LOG_FILE - 1;
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Can you drop the -1 and just use less than, instead of less than or equal. I think its more clear that way.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 3]
```diff
@@ -1064,13 +1061,7 @@ public void append(List<LogData> range) {
 
         // Extract all addresses associated with the provided write range.
         Set<Long> pendingWrites = range.stream()
-                .map(entry -> entry.getGlobalAddress()).collect(Collectors.toSet());
-
-        // Input validation.
-        if (pendingWrites.size() != range.size()) {
-            log.error("Input validation failed! Received entries are not unique.");
-            throw new OverwriteException(OverwriteCause.SAME_DATA);
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Client side also does this validation

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 4]
```diff
@@ -1195,6 +1186,7 @@ private void closeSegmentHandlers(long endSegment) {
                         sh.getSegment(), sh.getRefCount()
                 );
             }
+            channelsToSync.remove(sh.getWriteChannel());
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Why is this needed? `sh.close();` will flush internal channels.

File: test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java [hunk 5]
```diff
@@ -450,25 +449,24 @@ public void testGetGlobalTail() {
...
+        assertThat((long) logs.list().length).isEqualTo(numSegments);
 
         final long endSegment = 25;
         long trimAddress = endSegment * StreamLogFiles.RECORDS_PER_LOG_FILE + 1;
 
         // Get references to the segments that will be trimmed
-        Set<SegmentHandle> trimmedHandles = new HashSet();
-        for (SegmentHandle sh : ((StreamLogFiles)log).getSegmentHandles()) {
+        Set<SegmentHandle> trimmedHandles = new HashSet<>();
+        for (SegmentHandle sh : log.getSegmentHandles()) {
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java]: Maybe rename this to `OpenSegments` ?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 2 above)
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: You can refactor this into a `getEndSegment` method, it seems like its being used in multiple places

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: I think its better to use getStartingSegment and keep the - 1, also make sure that you don't close the first active segment in `closeSegmentHandlers` cause an unnecessary reload.

Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Done, changed to getStartingSegment

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 4 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: This is to ensure that next time when sync() is called by batchProcessor, there are no closed channels in the channelsToSync map.

File: test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java (hunk 5 above)
Comment (WenbinZhu) [reply test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 2 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: changed to use getStartingSegment. In my new GC patch #1921 , I extracted the linear addressing method: address / RECORDS_PER_LOG_FILE  and cleaned up all usages of it.
